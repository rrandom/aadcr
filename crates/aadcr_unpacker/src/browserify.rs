use indexmap::{map::Entry, IndexMap};

use oxc_allocator::{Allocator, CloneIn};
use oxc_ast::{
    ast::{
        Argument, ArrayExpressionElement, CallExpression, Expression, ObjectPropertyKind, Program,
        PropertyKey,
    },
    AstBuilder, AstKind,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_semantic::SemanticBuilder;
use oxc_span::SPAN;

use crate::{common::fun_to_program::FunctionToProgram, Module};

use super::{UnpackResult, UnpackReturn};

pub fn get_modules_form_browserify<'a>(
    allocator: &'a Allocator,
    program: &Program<'a>,
) -> UnpackResult<'a> {
    let ast = AstBuilder::new(allocator);

    let semantic = SemanticBuilder::new("").build(program).semantic;

    let nodes = semantic.nodes();
    let program_source_type = nodes
        .root()
        .map(|id| nodes.get_node(id).kind().as_program().unwrap().source_type);

    let program_directives = nodes.root().map(|id| {
        nodes
            .get_node(id)
            .kind()
            .as_program()
            .unwrap()
            .directives
            .clone_in(allocator)
    });

    let mut module_map = IndexMap::new();

    let mut entry_ids = vec![];

    let mut module_mapping = IndexMap::new();

    let mut errors = vec![];

    for node in nodes.iter() {
        let AstKind::CallExpression(CallExpression {
            callee: _,
            arguments,
            ..
        }) = node.kind()
        else {
            continue;
        };
        let [Argument::ObjectExpression(modules_obj), Argument::ObjectExpression(_), Argument::ArrayExpression(entry_id_array)] =
            arguments.as_slice()
        else {
            continue;
        };

        if !entry_id_array
            .elements
            .iter()
            .all(|ele| matches!(ele, ArrayExpressionElement::NumericLiteral(_)))
        {
            continue;
        }

        for ele in entry_id_array.elements.iter() {
            let ArrayExpressionElement::NumericLiteral(num) = ele else {
                unreachable!();
            };
            entry_ids.push(num.raw);
        }

        for property in modules_obj.properties.iter() {
            let ObjectPropertyKind::ObjectProperty(prop) = property else {
                break;
            };
            let PropertyKey::NumericLiteral(module_id) = &prop.key else {
                break;
            };

            let Expression::ArrayExpression(arr) = &prop.value else {
                break;
            };

            let [module_factory, ArrayExpressionElement::ObjectExpression(map)] =
                arr.elements.as_slice()
            else {
                break;
            };

            let is_module_map = map.properties.iter().all(|p| {
                let ObjectPropertyKind::ObjectProperty(prop) = p else {
                    return false;
                };
                matches!(prop.key, PropertyKey::StringLiteral(_))
                    && matches!(prop.value, Expression::NumericLiteral(_))
            });
            if !is_module_map {
                break;
            }

            for property in map.properties.iter() {
                let ObjectPropertyKind::ObjectProperty(prop) = property else {
                    unreachable!()
                };

                let PropertyKey::StringLiteral(short_name) = &prop.key else {
                    unreachable!()
                };

                let Expression::NumericLiteral(module_id) = &prop.value else {
                    unreachable!()
                };
                let short_name = short_name.value.as_str();

                let entry = module_mapping.entry(module_id.raw);
                match entry {
                    Entry::Occupied(mut e) if *e.get() != short_name => {
                        errors.push(OxcDiagnostic::warn(format!(
                            "Module {} has multiple short names: {} and {}",
                            module_id.raw,
                            e.get(),
                            short_name
                        )));
                        e.insert(short_name);
                    }
                    Entry::Vacant(e) => {
                        e.insert(short_name);
                    }
                    _ => {}
                }
            }

            let Some(module_factory) = module_factory.as_expression() else {
                break;
            };

            module_map.insert(module_id.raw, module_factory);
        }
    }
    if module_map.is_empty() {
        return None;
    }

    let mut modules = vec![];
    for (module_id, expr) in module_map {
        let fun_statement = ast.statement_expression(SPAN, expr.clone_in(allocator));

        let directives = program_directives
            .clone_in(allocator)
            .unwrap_or_else(|| ast.vec());

        let mut program = ast.program(
            SPAN,
            program_source_type.unwrap().clone_in(allocator),
            None,
            directives,
            ast.vec1(fun_statement),
        );

        let mut fun_renamer = FunctionToProgram::new(allocator, ["require", "module", "exports"]);
        fun_renamer.build(&mut program);

        modules.push(Module::new(
            module_id.to_string(),
            entry_ids.contains(&module_id),
            program,
        ));
    }
    Some(UnpackReturn {
        modules,
        module_mapping: Some(module_mapping),
        errors,
    })
}
