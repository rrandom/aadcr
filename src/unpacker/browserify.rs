use indexmap::IndexMap;

use oxc_allocator::{Allocator, CloneIn};
use oxc_ast::{
    ast::{
        Argument, ArrayExpressionElement, CallExpression, Expression, ObjectPropertyKind, Program,
        PropertyKey,
    },
    AstBuilder, AstKind,
};
use oxc_semantic::SemanticBuilder;
use oxc_span::Span;

use crate::unpacker::{common::fun_to_program::FunctionToProgram, Module};

pub fn get_modules_form_browserify<'a>(
    allocator: &'a Allocator,
    program: &Program<'a>,
) -> Option<std::vec::Vec<Module<'a>>> {
    let ast = AstBuilder::new(allocator);

    let semantic = SemanticBuilder::new("").build(program).semantic;

    let nodes = semantic.nodes();
    let program_source_type = nodes
        .root()
        .map(|id| nodes.get_node(id).kind().as_program().unwrap().source_type);

    let program_directives = nodes.root().map(|id| {
        nodes
            .get_node(id)
            .kind()
            .as_program()
            .unwrap()
            .directives
            .clone_in(allocator)
    });

    let mut module_map = IndexMap::new();

    let mut entry_ids = vec![];

    for node in nodes.iter() {
        let AstKind::CallExpression(CallExpression {
            callee: _,
            arguments,
            ..
        }) = node.kind()
        else {
            continue;
        };
        let [Argument::ObjectExpression(modules_obj), Argument::ObjectExpression(_), Argument::ArrayExpression(entry_id_array)] =
            arguments.as_slice()
        else {
            continue;
        };

        if !entry_id_array
            .elements
            .iter()
            .all(|ele| matches!(ele, ArrayExpressionElement::NumericLiteral(_)))
        {
            continue;
        }

        for ele in entry_id_array.elements.iter() {
            let ArrayExpressionElement::NumericLiteral(num) = ele else {
                unreachable!();
            };
            entry_ids.push(num.raw);
        }

        for property in modules_obj.properties.iter() {
            let ObjectPropertyKind::ObjectProperty(prop) = property else {
                break;
            };
            let PropertyKey::NumericLiteral(module_id) = &prop.key else {
                break;
            };

            let Expression::ArrayExpression(arr) = &prop.value else {
                break;
            };

            let [module_factory, ArrayExpressionElement::ObjectExpression(map)] =
                arr.elements.as_slice()
            else {
                break;
            };

            let is_module_map = map.properties.iter().all(|p| {
                let ObjectPropertyKind::ObjectProperty(prop) = p else {
                    return false;
                };
                matches!(prop.key, PropertyKey::StringLiteral(_))
                    && matches!(prop.value, Expression::NumericLiteral(_))
            });
            if !is_module_map {
                break;
            }

            let Some(module_factory) = module_factory.as_expression() else {
                break;
            };

            module_map.insert(module_id.raw, module_factory);
        }
    }
    if module_map.is_empty() {
        return None;
    }

    let mut modules = vec![];
    for (module_id, expr) in module_map {
        let fun_statement = ast.statement_expression(Span::default(), expr.clone_in(allocator));

        let directives = program_directives
            .clone_in(allocator)
            .unwrap_or_else(|| ast.vec());

        let mut program = ast.program(
            Span::default(),
            program_source_type.unwrap().clone_in(allocator),
            None,
            directives,
            ast.vec1(fun_statement),
        );

        let mut fun_renamer = FunctionToProgram::new(allocator, ["require", "module", "exports"]);
        fun_renamer.build(&mut program);

        modules.push(Module::new(
            module_id.to_string(),
            entry_ids.contains(&module_id),
            program,
        ));
    }
    Some(modules)
}
