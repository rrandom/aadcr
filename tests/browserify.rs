use std::path::Path;

use aadcr::unpacker::browserify::get_modules_form_browserify;
use oxc_allocator::Allocator;
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;
use oxc_span::SourceType;

#[test]
fn browserify() {
    let name = "tests/fixtures/browserify.js";
    let path = Path::new(&name);
    let source_text = std::fs::read_to_string(path).expect("{name} not found");

    let allocator = Allocator::default();
    let source_type = SourceType::from_path(path).unwrap();

    let ret = Parser::new(&allocator, &source_text, source_type).parse();

    assert!(ret.errors.is_empty());

    let program = ret.program;

    let ret = SemanticBuilder::new(&source_text)
        .with_excess_capacity(2.0)
        .build(&program);

    assert!(ret.errors.is_empty());

    let unpack_result = get_modules_form_browserify(&allocator, &program);

    assert!(unpack_result.is_some());

    let mut result = unpack_result.unwrap();

    assert_eq!(result.modules.len(), 4);

    insta::with_settings!({ prepend_module_to_snapshot => false, omit_expression => true }, {
      insta::assert_debug_snapshot!("browserify", result
        .modules
        .iter_mut()
        .map(|m| {
            m.write_code();
            m
        })
        .collect::<Vec<_>>()
    );
    });
}
