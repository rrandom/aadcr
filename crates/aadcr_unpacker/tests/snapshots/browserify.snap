---
source: tests/browserify.rs
assertion_line: 39
snapshot_kind: text
---
[
    Module {id: 1, is_entry: false, code: `function add(a, b) {
    	return a + b;
    }
    module.exports = add;
    ` },
    Module {id: 2, is_entry: true, code: `var sum = require("./sum");
    var numbers = [
    	1,
    	2,
    	3
    ];
    var result = sum(numbers);
    var outputElement = document.getElementById("output");
    outputElement.innerHTML = result;
    ` },
    Module {id: 3, is_entry: false, code: `function reduce(list, iteratee, memo) {
    	list.forEach(function(item) {
    		memo = iteratee(item, memo);
    	});
    	return memo;
    }
    module.exports = reduce;
    ` },
    Module {id: 4, is_entry: false, code: `var reduce = require("./reduce");
    var add = require("./add");
    function sum(list) {
    	return reduce(list, add, 0);
    }
    module.exports = sum;
    ` },
]
